Autolink declarations

esad opened the pull request:
This solves https://github.com/realm/jazzy/issues/483 - autolinking return and argument types in the method declarations in both Swift and Objective-C.

I've reused the same autolinking logic already used for text, with an additional flag that instead of running it inside `<code>`, runs it inside rogue (syntax highlighting) emitted `<span>` tags of special classes.

The diff for the integration specs will be quite huge as this affects lots of generated html - so it would be a good idea if you someone can take a look at what this does to their own docs - the output for both Realm ObjC and Swift looks good (and has nice links in the return and argument types!)

esad commented on lib/jazzy/sourcekitten.rb:
```diff
@@ -448,6 +458,7 @@ def self.autolink(docs, root_decls)
         doc.abstract = autolink_text(doc.abstract, doc, root_decls)
         doc.return = autolink_text(doc.return, doc, root_decls) if doc.return
         doc.children = autolink(doc.children, root_decls)
+        doc.declaration &&= autolink_text(doc.declaration, doc, root_decls, true)
```
Ok perhaps this could be rewritten to be consisent with the above `doc.return = autolink(..) if doc.return` style?

esad replied:
If I refactor this to `doc.declaration = .. if doc.declaration` the line gets too long and rubocop complains :( Should I refactor anyway and disable the rubocop warning?

segiddins commented on lib/jazzy/sourcekitten.rb:
```diff
@@ -418,10 +418,20 @@ def self.name_traversal(name_parts, doc)
+    def self.autolink_text(text, doc, root_decls, highlighted = false)
+      opening_tag_re, closing_tag_re = 
+        if highlighted
+          [%r{<span class="(?:n|kt)">},'</span>']
```
missing spaces between array elements

jpsim commented:
Woohoo! This is nice. Could you please address @segiddins's style feedback, push the changes to the integration specs (you should have commit access now :tada:) and then we'll be able to take a look at the specs diff and look at the changes in a web browser.

Thanks!

jpsim commented:
@esad now that I've made a few fixes to jazzy and generally brought it back to good health (or at least I hope), could you please revive this so we can include it in the next release?

esad commented:
@jpsim Almost there. Rubocop is complaining about a few things, so I wanted to ask for advice how to best deal with it:

<code>
lib/jazzy/sourcekitten.rb:440:71: C: Method has too many lines. [24/20]
lib/jazzy/sourcekitten.rb:440:71: C: Perceived complexity for autolink_text is too high. [8/7]
lib/jazzy/sourcekitten.rb:480:70: C: Line is too long. [98/80]
        doc.declaration = autolink_text(doc.declaration, doc, root_decls, true) if doc.declaration
</code>

 Not sure if I can split the `autolink_text` method meaningfully to avoid complexity and too many lines errors. Can I just disable Rubocop here?

Regarding the line length offence for `doc.declaration` - should we switch all three invocations to use `&&=` syntax?

jpsim commented:
This is a great start @esad! There are a few behaviors that I think should be addressed before getting this in:

First, disabling auto-linking to the same page you're already on:

![image](https://cloud.githubusercontent.com/assets/474794/14332825/330ac7f8-fc00-11e5-93d8-cf8601f3fc2c.png)

This is a screenshot taken from building the integration specs with this branch, and includes links from Alamofire's `Manager` to `Manager`.

Second, disabling links to types not defined in the module:

![image](https://cloud.githubusercontent.com/assets/474794/14332912/bfc305de-fc00-11e5-8a3d-e67d048eb1a4.png)

This is another screenshot taken from building the integration specs with this branch, and includes links from Alamofire's `SessionDelegate` to `NSURLRequest` because Alamofire contains extensions to `NSURLRequest`.

Those are the only two issues I can identify with this for now. Another modification we could consider is preserving the syntax color of the type from rouge, distinguishing between hyperlinks and non-links some other way, like underlines... I'm not too picky if we do this or not.

jpsim commented:
As for the rubocop violations, do whatever you prefer there.

DangerCI commented:
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-kind="Error">
        
            1 Error
        
      </th>
     </tr>
  </thead>
  <tbody>
        <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="true"><p>Please include a CHANGELOG entry. 
You can find it at <a href="https://github.com/realm/jazzy/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</p>
</td>
    </tr>
    
    
  </tbody>
</table>
  

  
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-kind="Warning">
        
            1 Warning
        
      </th>
     </tr>
  </thead>
  <tbody>
        <tr>
      <td>:warning:</td>
      <td data-sticky="true">This PR may need tests.</td>
    </tr>
    
    
  </tbody>
</table>
  

  
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-kind="Message">
        
            1 Message
        
      </th>
     </tr>
  </thead>
  <tbody>
        <tr>
      <td>:book:</td>
      <td data-sticky="true">Note, we hard-wrap at 80 chars and use 2 spaces after the last line.</td>
    </tr>
    
    
  </tbody>
</table>
  




<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://github.com/danger/danger/">danger</a>
</p>

esad commented:
@jpsim

I've addressed the two things you addressed above, updated the integration specs and tweaked rubocop rules a bit. Travis is failing because of missing CHANGELOG entry, I was think that you may want to craft that one?

The fact that we don't autolink on the same page anymore will also affect previous autolinking logic, where this was not respected, so for example in the abstract for https://realm.io/docs/objc/latest/api/Classes/RLMRealm.html#/c:objc(cs)RLMRealm(cm)defaultRealm, the `RLMRealm` won't be autolinked.

I've also updated the logic not to perform autolinking on extensions.

Furthermore, I had some issues with integration specs for MiscJazzyFeatures having extra `*.md` files reside in a `Docs/` subdirectory where also the autogenerated docs go per default, so I moved those to `Extra/*.md`.

As for Rubocop, I've tried to disable as little checks as possible and comply with the style, but I had to disable complexity checks for the autolinking method, as I think it's still quite readable. I've also globally disabled a recently introduce style check that IMHO makes no sense (see https://github.com/bbatsov/rubocop/issues/2338), which was also causing some old code in `sourcekitten.rb` to fail the check.

I've also updated the code to autolink declarations from the the "other language declaration", introduced in v0.6.1.

jpsim commented:
Hey @esad, there were conflicts since merging #569 so I worked to resolve those and rebuild the integration specs to better review this.

I did this in the [jazzy@jp-autolink-decls](https://github.com/realm/jazzy/compare/jp-autolink-decls) and [specs@jp-autolink-decls](https://github.com/realm/jazzy-integration-specs/compare/jp-autolink-decls) branches.

The main issue I can find with this are that some declarations now link to themselves:

![image](https://cloud.githubusercontent.com/assets/474794/15914402/7bb98edc-2d95-11e6-9492-5cb9a677d704.png)

esad commented:
@jpsim I've merged your changes and extended the check to prevent autolinking to the same page in the cases like the one you mentioned above. Can you have another look?

jpsim commented:
I'll review this again today @esad. I have to thank you for your patience, I know it's been tedious to keep this updated 😬

segiddins commented:
CI is only failing due to lack of a CHANGELOG entry

jpsim commented:
Previously, we'd link to other child declarations on the current page, like properties on class documentation. These no longer autolink

**before:**

![image](https://cloud.githubusercontent.com/assets/474794/15938137/c15940ea-2e26-11e6-9db6-531e002ab9eb.png)

**after:**

![image](https://cloud.githubusercontent.com/assets/474794/15938164/dfa5631c-2e26-11e6-9324-a33c65589a04.png)

esad commented:
@jpsim I thought that's what the intended behavior per your comment in https://github.com/realm/jazzy/pull/490#issuecomment-206569195 was - can you clarify in more detail what do you think the behavior should be when it comes to linking to the same page? Link everything except to itself and to the toplevel declaration of the current page (`Manager` in your example?)

jpsim commented:
Sorry for not being clear, yes I agree with your description of the intended behavior:

> Link everything except to itself and to the top-level declaration of the current page

esad commented:
@jpsim Updated. Could you have one more look?

jpsim commented:
Looks good, I've continued this in #588

esad commented:
@jpsim Thanks for taking over with the latest merge, looks good.
